dotnet/src/SemanticKernel.UnitTests/Functions/KernelFunctionFromMethodTests1.cs

```diff
@@ -1204,4 +1207,129 @@ public async Task ItPropagatesOriginalExceptionTypeAsync()
+        Assert.NotNull(actualArgValue);
+        Assert.Equal(28, actualArgValue.Id);
+    }
+
+    [Fact]
+    public async Task ItCanDeserializeJsonNodeAsync()
+    {
+        // Arrange
+        var node = JsonNode.Parse(@"{""id"":28}");
+        CustomTypeForJsonTests? actualArgValue = null;
+
+        var func = KernelFunctionFactory.CreateFromMethod((CustomTypeForJsonTests param) => { actualArgValue = param; });
+
+        // Act
+        var res = await func.InvokeAsync(this._kernel, new() { ["param"] = node });
+
+        // Assert
+        Assert.NotNull(actualArgValue);
+        Assert.Equal(28, actualArgValue.Id);
+    }
+
+    [Fact]
+    public async Task ItShouldNotDeserializeIfParameterTypeAndArgumentTypeAreSameAsync()
+    {
+        // Arrange
+        var node = JsonNode.Parse(@"{""id"":28}");
+        JsonNode? actualArgValue = null;
+
+        var func = KernelFunctionFactory.CreateFromMethod((JsonNode? param) => { actualArgValue = param; });
+
+        // Act
+        var res = await func.InvokeAsync(this._kernel, new() { ["param"] = node });
+
+        // Assert
+        Assert.NotNull(actualArgValue);
+        Assert.Same(node, actualArgValue);
+    }
+
+    [Fact]
+    public async Task ItCanDeserializeJsonStringAsync()
+    {
+        // Arrange
+        var jsonString = @"{""id"":28}";
+        CustomTypeForJsonTests? actualArgValue = null;
+
+        var func = KernelFunctionFactory.CreateFromMethod((CustomTypeForJsonTests param) => { actualArgValue = param; });
+
+        // Act
+        var res = await func.InvokeAsync(this._kernel, new() { ["param"] = jsonString });
+
+        // Assert
+        Assert.NotNull(actualArgValue);
+        Assert.Equal(28, actualArgValue.Id);
+    }
+
+    [Fact]
+    public async Task ItCanDeserializeThirdPartyJsonPrimitivesAsync()
+    {
+        // Arrange
+        var thirdPartyJsonPrimitive = new ThirdPartyJsonPrimitive(@"{""id"":28}");
+        CustomTypeForJsonTests? actualArgValue = null;
+
+        var func = KernelFunctionFactory.CreateFromMethod((CustomTypeForJsonTests param) => { actualArgValue = param; });
+
+        // Act
+        var res = await func.InvokeAsync(this._kernel, new() { ["param"] = thirdPartyJsonPrimitive });
+
+        // Assert
+        Assert.NotNull(actualArgValue);
+        Assert.Equal(28, actualArgValue.Id);
+    }
+
+#pragma warning disable CA1812 // Avoid uninstantiated internal classes
+    private sealed class CustomTypeForJsonTests
+#pragma warning restore CA1812 // Avoid uninstantiated internal classes
+    {
+        [JsonPropertyName("id")]
+        public int Id { get; set; }
+    }
+
+    private sealed class ThirdPartyJsonPrimitive
+    {
+        private readonly string _jsonToReturn;
+
+        public ThirdPartyJsonPrimitive(string jsonToReturn)
+        {
+            this._jsonToReturn = jsonToReturn;
+        }
+
+        public override string ToString()
+        {
+            return this._jsonToReturn;
+        }
+    }
```

Comment:
Are we able to come up with a function calling example where the LLM will provide a complex argument that can now be parsed successfully? That'd help to validate the full roundtrip, from .NET type to JSON schema to describe the parameter to JSON response for the parameter to deserialized .NET instance.